ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/MetricSink.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.parse.repl.metric.event.ReplicationMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * MetricSink.
+ * Scheduled thread to poll from Metric Collector and persists to DB
+ */
+public final class MetricSink {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricSink.class);
+  private ScheduledExecutorService executorService;
+  private static volatile MetricSink instance;
+  private boolean isInitialised = false;
+  private HiveConf conf;
+
+  private MetricSink() {
+    this.executorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  public static MetricSink getInstance() {
+    if (instance == null) {
+      synchronized (MetricSink.class) {
+        if (instance == null) {
+          instance = new MetricSink();
+        }
+      }
+    }
+    return instance;
+  }
+
+  public synchronized void init(HiveConf conf) {
+    if (!isInitialised) {
+      this.conf = conf;
+      long frequencyInMins = MetastoreConf.getLongVar(conf, MetastoreConf.ConfVars.REPL_METRICS_UPDATE_FREQUENCY);
+      this.executorService.schedule(new MetricSinkWriter(conf), frequencyInMins, TimeUnit.MINUTES);
+      isInitialised = true;
+    }
+  }
+
+  static class MetricSinkWriter implements Runnable {
+    private MetricCollector collector;
+    private HiveConf conf;
+
+    // writer instance
+
+    MetricSinkWriter(HiveConf conf) {
+      this.collector = MetricCollector.getInstance();
+      this.conf = conf;
+    }
+
+    @Override
+    public void run() {
+      // get metrics
+      List<ReplicationMetric> metrics = collector.getMetrics();
+      // write metrics and retry if fails
+      Retry<Void> retriable = new Retry<Void>(Exception.class) {
+        @Override
+        public Void execute() throws Exception {
+          ReplicationMetricList metricList = new ReplicationMetricList();
+          List<ReplicationMetrics> replicationMetricsList = new ArrayList<>();
+          for (ReplicationMetric metric : metrics) {
+            ReplicationMetrics persistentMetric = new ReplicationMetrics();
+            persistentMetric.setDumpExecutionId(metric.getDumpExecutionId());
+            persistentMetric.setScheduledExecutionId(metric.getScheduledExecutionId());
+            persistentMetric.setPolicy(metric.getPolicy());
+            ObjectMapper mapper = new ObjectMapper();
+            persistentMetric.setProgress(mapper.writeValueAsString(metric.getProgress()));
+            persistentMetric.setMetadata(mapper.writeValueAsString(metric.getMetadata()));
+            replicationMetricsList.add(persistentMetric);
+          }
+          metricList.setReplicationMetricList(replicationMetricsList);
```

Comment:
At this point of time, we have  2X of metric data we are trying  to persist. Can we use some other DS, like linkedList so, that we can remove the ReplicationMetric from original list  while trying to get and add a corresponding ReplicationMetrics?

Reply:
This is just an assignment. Where is the 2X memory coming from? The list in metricList is not allocated any extra memory apart from the list size.

Reply:
And from original list you can't modify or remove as retry is based on that. We don't want the retry to be broken. If we remove all elements from the first list and add to persistent list and then retry fails, how will that be handled.

Reply:
we have two lists here metrics and replicationMetricsList, we are getting one ReplicationMetric from metrics, forming an equivalent thrift compatible object and adding it to another list(replicationMetricsList). At the end of for loop, we will have 2X of data (from two lists of same size).
Retry problem: We have retry at wrong place , ideally only "Hive.get(conf).getMSC().addReplicationMetrics(metricList); should be retried, as copying is not the error prone part we are trying to retry.

Reply:
However, even if we remove objects from metrics, it actually will not be subject to  garbage collection as the same object is being referenced from ReplicationMetricCollector.
So we should just fix the concurrency issue in ReplicationMetric object.

Reply:
Coping may also fail. Better to retry both.